src/json-api/json-api-hapi-plugin.ts

```diff
@@ -589,20 +640,69 @@ export class JsonApiHapiPlugin {
     return reply(this.getEntity('notification', api => api.getProjectNotificationConfigurations(projectId)));
   }
 
-  public async postNotificationConfigurationHandler(request: Hapi.Request, reply: Hapi.IReply) {
-    const config = request.payload.data.attributes;
-
-    config.teamId = config.teamId || null;
-    config.projectId = config.projectId || null;
+  private async authorizeNotificationConfiguration(request: Hapi.Request, reply: Hapi.IReply) {
+    try {
+      reply(request.payload.data.attributes);
+    } catch (error) {
+      reply(Boom.badRequest());
+    }
+  }
 
-    if (!config.teamId && !config.projectId) {
-      throw Boom.badRequest('teamId or projectId should be defined');
+  private async authorizeTeamOrProjectAccess(request: Hapi.Request, reply: Hapi.IReply) {
+    try {
+      const config = request.pre.filter;
+
+      config.teamId = config.teamId || null;
+      config.projectId = config.projectId || null;
+
+      if (!config.teamId && !config.projectId) {
+        return reply(Boom.badRequest('teamId or projectId should be defined'));
+      }
+      if (config.teamId && config.projectId) {
+        return reply(Boom.badRequest('teamId and projectId should not both be defined'));
+      }
+      if (config.projectId) {
+        const project = await request.gitlab.getProject(config.projectId, request.auth.credentials.username);
```

Comment:
We can then look at memoizing it, etc.

You could pass the function along the request, so that we avoid referencing `gitlab` at all here. I would prefer that knowledge of us using gitlab is encapsulated mainly to the project and deployment modules.